shim/shim.c

```diff
@@ -0,0 +1,631 @@
+		shim->exiting = true;
+		free(buf);
+		return;
+	} else if (shim->exiting && stream_len == (STREAM_HEADER_SIZE+1)) {
+		code = atoi(buf + STREAM_HEADER_SIZE); 	// hyperstart has sent the exit status
+		free(buf);
+		exit(code);
+	}
+
+	/* TODO: what if writing to stdout/err blocks? Add this to the poll loop
+	 * to watch out for EPOLLOUT
+	 */
+	offset = STREAM_HEADER_SIZE;
+	while (offset < stream_len) {
+		ret = write(outfd, buf+offset, (size_t)(stream_len-offset));
+		if (ret <= 0 ) {
+			free(buf);
+			return;
+		}
+		offset += (ssize_t)ret;
+	}
+	free(buf);
+}
+
+/*!
+ * Handle data on the proxy ctl socket fd
+ *
+ *\param shim \ref cc_shim
+ */
+void
+handle_proxy_ctl(struct cc_shim *shim)
+{
+	char buf[LINE_MAX];
+	ssize_t ret;
+
+	if (! shim) {
+		return;
+	}
+
+	ret = read(shim->proxy_sock_fd, buf, LINE_MAX);
+	if (ret == -1) {
+		shim_warning("Error reading from the proxy ctl socket: %s\n", strerror(errno));
+		exit(EXIT_FAILURE);
+	} else if (ret == 0) {
+		shim_warning("EOF received on proxy ctl socket. Proxy has exited\n");
+		exit(EXIT_FAILURE);
+	}
+
+	//TODO: Parse the json and log error responses explicitly
+	shim_debug("Proxy response:%s\n", buf + PROXY_CTL_HEADER_SIZE);
+}
+
+long long
+parse_numeric_option(char *input) {
+	char       *endptr;
+	long long   num;
+
+	errno = 0;
+	num = strtoll(input, &endptr, 10);
+	if ( errno || *endptr ) {
+		return -1;
+	}
+	return num;
+}
+
+/*!
+ * Print program usage
+ */
+void
+print_usage(void) {
+	printf("%s: Usage\n", program_name);
+        printf("  -c,  --container-id   Container id\n");
+        printf("  -p,  --proxy-sock-fd  File descriptor of the socket connected to cc-proxy\n");
+        printf("  -o,  --proxy-io-fd    File descriptor of I/0 fd sent by the cc-proxy\n");
+        printf("  -s,  --seq-no         Sequence no for stdin and stdout\n");
+        printf("  -e,  --err-seq-no     Sequence no for stderr\n");
+        printf("  -d,  --debug          Enable debug output\n");
+        printf("  -h,  --help           Display this help message\n");
+}
+
+int
+main(int argc, char **argv)
+{
+	struct cc_shim shim = {
+		.container_id   =  NULL,
+		.proxy_sock_fd  = -1,
+		.proxy_io_fd    = -1,
+		.io_seq_no      =  0,
+		.err_seq_no     =  0,
+		.exiting        =  false,
+	};
+	struct pollfd      poll_fds[MAX_POLL_FDS] = {{-1}};
+	nfds_t             nfds = 0;
+	int                ret;
+	struct sigaction   sa;
+	int                c;
+	bool               debug = false;
+
+	program_name = argv[0];
+
+	struct option prog_opts[] = {
+		{"container-id", required_argument, 0, 'c'},
+		{"proxy-sock-fd", required_argument, 0, 'p'},
+		{"proxy-io-fd", required_argument, 0, 'o'},
+		{"seq-no", required_argument, 0, 's'},
+		{"err-seq-no", required_argument, 0, 'e'},
+		{"debug", no_argument, 0, 'd'},
+		{"help", no_argument, 0, 'h'},
+		{ 0, 0, 0, 0},
+	};
+
+	while ((c = getopt_long(argc, argv, "c:p:o:s:e:dh", prog_opts, NULL))!= -1) {
```

Comment:
Super pedantic I know, but I always sort the option string and the case statements for easier maintenance.